service/frontend/workflow_handler.go

```diff
@@ -2743,27 +2747,59 @@ func (wh *WorkflowHandler) DescribeWorkflowExecution(ctx context.Context, reques
 		return nil, err
 	}
 
-	response, err := wh.historyClient.DescribeWorkflowExecution(ctx, &historyservice.DescribeWorkflowExecutionRequest{
-		NamespaceId: namespaceID.String(),
-		Request:     request,
-	})
+	response, err := wh.historyClient.DescribeWorkflowExecution(
+		ctx,
+		&historyservice.DescribeWorkflowExecutionRequest{
+			NamespaceId: namespaceID.String(),
+			Request:     request,
+		},
+	)
 
 	if err != nil {
 		return nil, err
 	}
 
-	if response.GetWorkflowExecutionInfo().GetSearchAttributes() != nil {
-		saTypeMap, err := wh.saProvider.GetSearchAttributes(wh.config.ESIndexName, false)
-		if err != nil {
-			return nil, serviceerror.NewUnavailable(fmt.Sprintf(errUnableToGetSearchAttributesMessage, err))
-		}
-		searchattribute.ApplyTypeMap(response.GetWorkflowExecutionInfo().GetSearchAttributes(), saTypeMap)
-		aliasedSas, err := searchattribute.AliasFields(wh.saMapper, response.GetWorkflowExecutionInfo().GetSearchAttributes(), request.GetNamespace())
-		if err != nil {
-			return nil, err
+	executionInfo := response.GetWorkflowExecutionInfo()
+	if executionInfo != nil {
+		if response.CloseVisibilityTaskCompleted {
+			// If close visibility task has completed, then search attributes and memo
+			// were removed from mutable state, and we need to fetch from visibility.
+			visResponse, err := wh.visibilityMrg.GetWorkflowExecution(
```

Comment:
Just want to confirm my understanding, we can't do this in history because history does not allow read operation on visibility?

Reply:
At least at this moment, there's no visibility calls from history.

Reply:
Actually, the history handler can call visibility. I can move the logic there, it makes more sense.

Reply:
Does it mean you don't need to expose CloseVisibilityTaskCompleted in the describe workflow response?

Reply:
It would still expose. The response is built in the history engine, which doesn't have the visibility client.

Reply:
I am not sure I follow. If history service can read from visibility, then the describe response from history service should not include `CloseVisibilityTaskCompleted` flag I believe.

Reply:
History service returns `DescribeWorkflowExecutionResponse`, which is built in the `describeworkflow` API.
The code execution path is: frontend -> history handler -> history engine -> describe workflow API.
The history handler has the visibility client, but history engine does not have it.
So, the only way to avoid adding that field at all is if we add the visibility client to history handler, and pass it to `describeworkflow` API.

Reply:
I guess I never tried to you to move this logic into history service. I just want to confirm if currently (before the change) history service can/allowed to read from visibility. If not, I am totally fine with returning the flag and let frontend do the loading. 

But looks like now configurations are added to allow history service to perform read, which I personally think is a perfectly valid first use case for history to perform visibility read. We may want to confirm this change with rest of the team though. 

so, I guess both approaches makes sense to me.